mainapp/helpers.py

```diff
@@ -170,26 +172,63 @@ def get_book_dict(bookid_list):
     return rec_books_dict
 
 
-def combine_ids(cv_bookids, embedding_bookids, already_rated, recommendations=9):
+def combine_ids(tfidf_bookids, embedding_bookids, already_rated, recommendations=9):
     '''
         Returns best bookids combining both approaches
         Embedding - Top 6
-        CV - Top 3
+        TF IDF - Top 3
     '''
-    cv_bookids = list(cv_bookids.difference(already_rated))
-    top_3_cv = set(cv_bookids[:3])
+    tfidf_bookids = list(tfidf_bookids.difference(already_rated))
+    top_3_tfidf = set(tfidf_bookids[:3])
     embedding_bookids = embedding_bookids.difference(already_rated)
-    embedding_bookids = list(embedding_bookids.difference(top_3_cv))
-    top_3_cv = list(top_3_cv)
+    embedding_bookids = list(embedding_bookids.difference(top_3_tfidf))
+    top_3_tfidf = list(top_3_tfidf)
     top_6_embed = list(embedding_bookids[:6])
-    best_bookids = top_3_cv + top_6_embed
+    best_bookids = top_3_tfidf + top_6_embed
 
     if len(best_bookids) < recommendations:
         # If not enough recommendations
-        best_bookids = best_bookids + cv_bookids[3:(recommendations + 3 - len(best_bookids))]
+        two_n = (recommendations - len(best_bookids))
+        n1, n2 = math.ceil(two_n/2), math.floor(two_n/2)
+
+        # n1 number of books from remaining tf_idf dataset
+        best_bookids_tfidf = tfidf_bookids[3: (3*2)+n1]
+        best_bookids_tfidf = list(set(best_bookids_tfidf).difference(set(best_bookids)))[:n1]
+
+        # n2 number of books from list of top rated books of the most common genre among the books yet recommended
+        genre_recomm_bookids = most_common_genre_recommendations(best_bookids, already_rated, best_bookids_tfidf, n2)
+
+        # number of recommendations = len(best_bookids) + n1 + n2 = len(best_bookids) + two_n
+        best_bookids = best_bookids + best_bookids_tfidf + genre_recomm_bookids
     return best_bookids
 
 
+def most_common_genre_recommendations(best_bookids, already_rated, best_bookids_tfidf, n):
+    '''
+        Returns n top rated of the most_common_genre among all lists taken as input
+    '''
+    # Final list of bookids to be recommended
+    books = set(best_bookids+list(already_rated)+best_bookids_tfidf)
+
+    # Accumulation of all the genres listed in `books` variable
+    genre_frequency = []
+    for book in books:
+        genre_frequency.extend(df_book[df_book['book_id'] == book]['genre'].values[0].split(", "))
+
+    # The most common genre among the bookids in `books` variable
+    most_common_genre = Counter(genre_frequency).most_common(1)[0][0]
+
```

Comment:
@saaalik I could follow till here, you got the most common genre, could you elaborate on the code below and why is there a loop needed?

Reply:
genre_wise function will always return 2*n number of books, but if a case came up where the unique sets of bookids are less than n after removing the common bookids from `books` i.e. after 
```python
genre_recommendations = genre_recommendations.difference(books)    #Might return a list of length<n in rare test cases#
```, then the genre_wise function will be executed again, and the weight calculation method always randomizes the output, so the loop will execute till the returned value of genre_wise function is less than n

Reply:
`books`  will always have a unique set since we took care of that, so whatever remaining `n` books we take from whatever genre, just make it `2*n` and take a `difference` and take first `n` books, don't see why loop would be needed

Reply:
Removed the while loop, please review

Reply:
Why is the code below changed after you finished with the implementation earlier and were about to write the tests, it was correct then